build-system/tasks/performance/measure-documents.js

```diff
@@ -0,0 +1,201 @@
+          longTask.startTime > firstContentfulPaint &&
+          longTask.duration > longest
+        ) {
+          longest = longTask.duration;
+        }
+      });
+
+      return longest;
+    }
+
+    function getTimeToInteractive() {
+      return Date.now() - window.measureStarted;
+    }
+
+    return {
+      visible: getMetric('visible'),
+      firstPaint,
+      firstContentfulPaint,
+      largestContentfulPaint: window.largestContentfulPaint,
+      timeToInteractive: getTimeToInteractive(),
+      maxFirstInputDelay: getMaxFirstInputDelay(),
+      cumulativeLayoutShift: window.cumulativeLayoutShift,
+    };
+  });
+
+/**
+ * Writes measurements to ./results.json
+ *
+ * @param {string} url
+ * @param {string} version
+ * @param {*} metrics
+ */
+function writeMetrics(url, version, metrics) {
+  let results = {};
+
+  if (fs.existsSync(RESULTS_PATH)) {
+    results = JSON.parse(fs.readFileSync(RESULTS_PATH));
+  }
+
+  if (!results[url]) {
+    results[url] = {[CONTROL]: [], [EXPERIMENT]: []};
+  }
+
+  results[url][version].push(metrics);
+
+  fs.writeFileSync(RESULTS_PATH, JSON.stringify(results));
+}
+
+/**
+ * Opens Chrome, loads the URL from local file cache, and collects
+ * metrics for the specified URL and version
+ *
+ * @param {string} url
+ * @param {string} version "control" or "experiment"
+ * @param {Object} options
+ * @return {Promise}
+ */
+async function measureDocument(url, version, {headless}) {
+  const browser = await puppeteer.launch({
+    headless,
+    args: [
+      '--allow-file-access-from-files',
+      '--enable-blink-features=LayoutInstabilityAPI',
+    ],
+  });
+
+  const page = await browser.newPage();
+  await page.setCacheEnabled(false);
+  await setupMeasurement(page);
+
+  try {
+    await page.goto(`file:${urlToCachePath(url, version)}`, {
+      waitUntil: 'networkidle0',
+    });
+  } catch {
+    // site did not load
+    await browser.close();
+    return;
+  }
+
+  const metrics = await readMetrics(page);
+  writeMetrics(url, version, metrics);
+  await browser.close();
+}
+
+/**
+ * Loads cached local copies of the URLs in Chrome with Puppeteer and
+ * runs a script on the page to collect performance metrics. Saves
+ * performance metrics to results.json in this directory.
+ *
+ * @param {Array<string>} urls
+ * @param {{headless:boolean, runs:number}} options
+ * @return {Promise} Fulfills when all URLs have been measured
+ */
+async function measureDocuments(urls, {headless, runs}) {
+  try {
+    fs.unlinkSync(RESULTS_PATH);
+  } catch {} // file does not exist (first run)
+
+  // Make an array of tasks to be executed
+  const tasks = urls.flatMap(url =>
+    Array.from({length: runs}).flatMap(() => [
+      measureDocument.bind(null, url, CONTROL, {headless}),
+      measureDocument.bind(null, url, EXPERIMENT, {headless}),
+    ])
+  );
+
+  // Excecute the tasks serially
+  const [first, ...rest] = tasks;
+  await rest.reduce((prev, task) => prev.then(task), first());
```

Comment:
Do we need to execute serially?

Reply:
I got more consistent results from executing serially, and on my machine saw overall execution time increased with concurrency